### PR TITLE
Use the untrusted server certificate chain to create a valid certificate ID for OCSP_request

### PR DESCRIPTION
**Pull request by olszomal**
The `openssl s_server` command-line tool does not use the untrusted certificate chain for building OCSP requests.

Let's consider the following `openssl s_server` options:
```
-cert infile
	The certificate to use.
-cert_chain
	A file or URI of untrusted certificates to use when attempting to build the certificate chain related to the certificate specified via the -cert option.
-CAfile file
	A file of trusted certificates.
```

An issuer certificate is required to create an OCSP_CERTID structure. This ID is placed in the OCSP request.
The issuer certificate can be untrusted, although any issuer can be designated as a trusted root authority.

Start the OCSP responder:
```
openssl ocsp -port 19254 -index index.txt -rsigner ocsp.pem -CA intermediateCA.pem
```

The following doesn't work but it should:
```
openssl s_server -accept 4433 -cert server.pem -cert_chain intermediateCA.pem -status_verbose
Using default temp DH parameters
ACCEPT
cert_status: callback called
cert_status: AIA URL: http://127.0.0.1:19254/ocsp
cert_status: Can't retrieve issuer certificate.
(...)
```

The following workaround works, but it requires trusting an untrusted certificate:
```
openssl s_server -accept 4433 -cert server.pem -CAfile intermediateCA.pem -status_verbose
Using default temp DH parameters
ACCEPT
cert_status: callback called
cert_status: AIA URL: http://127.0.0.1:19254/ocsp
cert_status: ocsp response sent:
OCSP Response Data:
    OCSP Response Status: successful (0x0)
    Response Type: Basic OCSP Response
    Version: 1 (0x0)
    Responder Id: C = PL, O = Stunnel Developers, OU = Intermediate CA, CN = Intermediate CA
    Produced At: Sep 15 12:07:08 2023 GMT
    Responses:
    Certificate ID:
      Hash Algorithm: sha1
      Issuer Name Hash: 48BDDBF0D915E576EE31434E57A6FAA988FADEB2
      Issuer Key Hash: AF74BB9A5BBE5FAF4650E3D0F6FE4F49DB3DBADA
      Serial Number: 3168B44D3E67DA3B9D7A7B40E08793F616EEF669
    Cert Status: good
    This Update: Sep 15 12:07:08 2023 GMT
	(...)
```

This PR enables the untrusted server certificate chain to find the issuer certificate.
```
openssl s_server -accept 4433 -cert server.pem -cert_chain intermediateCA.pem -status_verbose 
Using default temp DH parameters
ACCEPT
cert_status: callback called
cert_status: AIA URL: http://127.0.0.1:19254/ocsp
cert_status: ocsp response sent:
OCSP Response Data:
    OCSP Response Status: successful (0x0)
    Response Type: Basic OCSP Response
    Version: 1 (0x0)
    Responder Id: C = PL, O = Stunnel Developers, OU = Intermediate CA, CN = Intermediate CA
    Produced At: Sep 15 12:08:46 2023 GMT
    Responses:
    Certificate ID:
      Hash Algorithm: sha1
      Issuer Name Hash: 48BDDBF0D915E576EE31434E57A6FAA988FADEB2
      Issuer Key Hash: AF74BB9A5BBE5FAF4650E3D0F6FE4F49DB3DBADA
      Serial Number: 3168B44D3E67DA3B9D7A7B40E08793F616EEF669
    Cert Status: good
    This Update: Sep 15 12:08:46 2023 GMT
	(...)
```

### CODE REVIEW

**Comment by mtrojnar** [comment]
@t8m Are you sure this was correctly triaged as a feature?  Without this PR, `openssl server` fails to build OCSP requests for non-trivial certificate chains.

**Comment by t8m** [comment]
Yeah, I stand by it. @openssl/otc anybody from OTC disagrees?

**Comment by t-j-h** [comment]
@t8m is correct in my view to see this as a feature request - we don't document that we use the cert_chain in OCSP contexts - if we did document that and it did not work it would be a defect - but I don't read the documentation (such as it is) as indicating that this should be expected to work.

Adding support for this makes sense and it would need to have some tests added so we ensure we don't inadvertently break it in subsequent releases.

**Comment by t-j-h** [comment]
@t8m why tests-exempted here - I'd expect some exercise of the new capability to be added as a test ...

**Comment by t8m** [comment]
@olszomal Would you be able to write a test?

**Comment by t8m** [comment]
> @t8m why tests-exempted here - I'd expect some exercise of the new capability to be added as a test ...

I assumed that for an external contributor it would be a quite non-trivial exercise to write a proper test recipe for this. I assumed it would take much more work than the bug fix itself. We allow exempting tests when disproportionate amount of work is needed for them. We can as well set `tests: deferred` and write the test ourselves - would for example @jamuir be willing to write it?

**Comment by mtrojnar** [comment]
> @olszomal Would you be able to write a test?

I only found one test involving s_server: https://github.com/openssl/openssl/blob/cf6342bc024868f5a55f2225f2e083415fb1329a/test/recipes/82-test_tfo_cli.t
I guess the best person to write another similar test would be someone who is already fluent in perl.

**Comment by jamuir** [comment]
> We can as well set `tests: deferred` and write the test ourselves - would for example @jamuir be willing to write it?

Yes, if needed I can write a test.

File: apps/s_server.c [hunk 1]
```diff
@@ -511,21 +513,31 @@ static int get_ocsp_resp_from_responder(SSL *s, tlsextstatusctx *srctx,
...
-        goto err;
-    obj = X509_STORE_CTX_get_obj_by_subject(inctx, X509_LU_X509,
-                                            X509_get_issuer_name(x));
-    if (obj == NULL) {
-        BIO_puts(bio_err, "cert_status: Can't retrieve issuer certificate.\n");
-        goto done;
+    if (chain) {
+        /* use untrusted server certificate chain (-cert_chain option) */
+        issuer = sk_X509_value(chain, 0);
+        if (!X509_name_cmp(X509_get_issuer_name(x), X509_get_subject_name(issuer)))
```
Comment (t8m) [inline apps/s_server.c:522]: Please use `X509_name_cmp() == 0`. We use boolean ops only on true/false return values.

**Review by t8m — CHANGES_REQUESTED** [review]
One nit

File: apps/s_server.c [hunk 2]
```diff
@@ -511,21 +513,31 @@ static int get_ocsp_resp_from_responder(SSL *s, tlsextstatusctx *srctx,
...
-    if (!X509_STORE_CTX_init(inctx,
-                             SSL_CTX_get_cert_store(SSL_get_SSL_CTX(s)),
-                             NULL, NULL))
-        goto err;
-    obj = X509_STORE_CTX_get_obj_by_subject(inctx, X509_LU_X509,
-                                            X509_get_issuer_name(x));
-    if (obj == NULL) {
-        BIO_puts(bio_err, "cert_status: Can't retrieve issuer certificate.\n");
-        goto done;
+    if (chain) {
```
Comment (t8m) [inline apps/s_server.c:519]: I am sorry for not noticing this before - this also needs to be `chain != NULL` according to our code style guide.

File: apps/s_server.c [hunk 3]
```diff
@@ -511,21 +513,31 @@ static int get_ocsp_resp_from_responder(SSL *s, tlsextstatusctx *srctx,
...
+        /* use untrusted server certificate chain (-cert_chain option) */
+        issuer = sk_X509_value(chain, 0);
+        if (X509_name_cmp(X509_get_issuer_name(x), X509_get_subject_name(issuer)) == 0)
```
Comment (t8m) [inline apps/s_server.c:522]: Could you please also split this line at the comma?

**Review by t8m — CHANGES_REQUESTED** [review]
Last two nits and I'll approve I promise.

**Review by t8m — APPROVED** [review]

**Comment by olszomal** [comment]
@jamuir In the [stunnel ](https://github.com/mtrojnar/stunnel/tree/master/tests/certs)repository on GitHub, there is a script that generates certificates along with configuration files. The certificate file names have been changed for the above example.
```
$ ./maketestcert.sh 
$ cp server_cert.pem server.pem
$ cp interCA_ocsp.pem ocsp.pem
```
Authority Info Access is set:
`authorityInfoAccess             = OCSP;URI:http://127.0.0.1:19254/ocsp`

**Comment by jamuir** [comment]
Thanks, @olszomal , for providing those files.

I have been able to reproduce the original issue and verify your fix.

When testing, there are three parties to consider:

```
tls client <---> tls server <---> ocsp responder 
```
I initiate the flow to `ocsp responder` from `tls client` using

```
openssl s_client -status ...
```
The flag `-status` is used to tell `tls server` that `tls client` wants ocsp data.

The tcp-fast-open test that mtrojnar pointed to (`82-test_tfo_cli.t`) checks messages written to the console by `s_server` and `s_client`.

I will do something similar, but possibly just for `s_server`.  We should see `cert_status: ocsp response sent:` and `OCSP Response Data:` written to the console by `s_server` when ocsp data is successfully relayed to `tls client`.

**Comment by jamuir** [comment]
see #23101 .

File: apps/s_server.c [hunk 4]
```diff
@@ -511,21 +513,32 @@ static int get_ocsp_resp_from_responder(SSL *s, tlsextstatusctx *srctx,
...
-                             NULL, NULL))
-        goto err;
-    obj = X509_STORE_CTX_get_obj_by_subject(inctx, X509_LU_X509,
-                                            X509_get_issuer_name(x));
-    if (obj == NULL) {
-        BIO_puts(bio_err, "cert_status: Can't retrieve issuer certificate.\n");
-        goto done;
+    if (chain != NULL) {
+        /* use untrusted server certificate chain (-cert_chain option) */
+        issuer = sk_X509_value(chain, 0);
```
Comment (mattcaswell) [inline apps/s_server.c:521]: Why are you only looking at the first cert in the stack? Surely the issuer could be anywhere in the stack?

**Review by mattcaswell — CHANGES_REQUESTED** [review]

**Review by t8m — APPROVED** [review]

**Comment by mtrojnar** [comment]
Do I correctly understand that this PR is ready for merging?

**Review by mattcaswell — CHANGES_REQUESTED** [review]
I might have expected to see some s_server documentation changes to explain how the intermediate certs affect this

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
ping @mattcaswell  for re-review

**Comment by t8m** [comment]
Ping for second review

**Comment by t8m** [comment]
Merged to the master branch. Thank you for your contribution.

**Comment by martinRa2** [comment]
Hello, 
I maintain the PR #20945 . I have integrated your changes into it. 

@olszomal it would be great if you could have a look if your code was successfully integrated into the PR.

**Pull request closed without merging**